pkg/querysharding/util.go

```diff
@@ -0,0 +1,76 @@
+package querysharding
+
+import (
+	"encoding/base64"
+	"sync"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+)
+
+const (
+	CortexShardByLabel = "__CORTEX_SHARD_BY__"
+)
+
+var (
+	buffers = sync.Pool{New: func() interface{} {
+		b := make([]byte, 0, 100)
+		return &b
+	}}
+)
+
+func InjectShardingInfo(query string, shardInfo *storepb.ShardInfo) (string, error) {
+	expr, err := parser.ParseExpr(query)
+	if err != nil {
+		return "", err
+	}
+	b, err := shardInfo.Marshal()
+	if err != nil {
+		return "", err
+	}
+	parser.Inspect(expr, func(n parser.Node, _ []parser.Node) error {
+		if selector, ok := n.(*parser.VectorSelector); ok {
+			selector.LabelMatchers = append(selector.LabelMatchers, &labels.Matcher{
+				Type:  labels.MatchEqual,
+				Name:  CortexShardByLabel,
+				Value: base64.StdEncoding.EncodeToString(b),
```

Comment:
I am not sure why we need to encode the serialized string? It makes the original string even longer, right? What's the benefit of it

Reply:
We are serializing protobuff so i was just afraid it may have non ASCII bytes and maybe break something on the way.

WDYT?

Reply:
I know we do have some label value limitations at the ingestion time. At the query time, I don't know that limitation though. I guess as long as the gRPC serialization/deserilization works then we should be good?

Reply:
Yeah.. seems that the value is not too big even encoding base64: `EAIYASIDcG9kIgxjbHVzdGVyX25hbWU=`

Reply:
Yeah, the length is short so probably fine. I am more thinking about debuggability and visibility. At tracing or some other tools if we need to look at the matchers then non-encoded version is pretty straightforward.

Reply:
This labels are sent on a `http.Request` as a string... hummmm I think is safer to encode this here..

And even if we dont base64 encode, the protobuff serialization will not be human readable right?

Reply:
I see. We are searilize it using protobuf rather than JSON. Then it is not readable anyway. I am fine to go with either encode or not as long as they work